### ISSUE TITLE
Docks appear on top of full screen windows

### ISSUE THREAD

**Issue by IvanMalison** [xmonad/xmonad-contrib#211]
### Problem Description

Docks/strut windows that are started AFTER a process with a window that later goes fullscreen appear on top of the fullscreen window. I was not able to determine if this issue is specific to chrome.

See: http://imgur.com/a/s55Bg
### Configuration File

https://gist.github.com/IvanMalison/ddfc7c474859f227ea101def10de3e4d
```haskell
module Main where

import XMonad
import XMonad.Hooks.EwmhDesktops
import XMonad.Hooks.ManageDocks

main =
  xmonad .
  docks .
  ewmh $ def { handleEventHook = handleEventHook def <+> fullscreenEventHook
             , layoutHook = avoidStruts $ layoutHook def
             }
```

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I tested my configuration with [xmonad-testing](https://github.com/xmonad/xmonad-testing)

**Comment by sgf-dma** [comment]
I think, this is expected: docks are unmanaged by xmonad (in `docks`) and `restackWindows`, which xmonad uses in `windows`, will not change window order relative to windows not listed in the call (i.e. unmanaged). That means, this issue should be reproducible with just toggling dock struts off: application window should go _under_ docks. @IvanMalison , please, check this.

To fix this we may call `lowerWindow` for all docks before unmanaging them. But this may have side-effect of docks going under desktop (e.g. nautilus). Or just restack docks below all managed windows before unmanaging them. I may write a PR, if there will be a decision how to deal with this. @geekosaur , @pjones any thoughts?

**Comment by de-vri-es** [comment]
I occasionally run into this as well if I need to restart my docks for some reason. Restacking them below managed windows seems like a good idea to me.

**Comment by dkasak** [comment]
I need this too. Does anyone have an easy workaround until this gets implemented?

**Comment by sgf-dma** [comment]
I've no time to finish this, sorry. I'll try to look at this again and may be remember how i wanted to fix that in 1-2 weeks.

**Comment by IvanMalison** [comment]
@sgf-dma Any thoughts on this?

**Comment by geekosaur** [comment]
Losing docks on restart suggests either they missed `docksStartupHook` or that that or its interaction with `avoidStruts` is buggy.

The original problem is more likely to be a bug in the dock; xmonad doesn't have a say in, or a way to communicate with, owners of struts. And while EWMH-style fullscreen is under xmonad's control, if something is using old-style (possibly because xmonad can't know when to add `_NET_WINDOW_STATE_FULLSCREEN` to `_NET_SUPPORTED`) it may put the fullscreen window in its original Z-order — and if it's `override_redirect`, xmonad won't know it was fullscreened.

There are also some potential problems from docks being completely self-managed vs. EWMH layers: even in window managers that support layers directly there is reliance on everything obeying layers that only the window manager knows the full story about, because docks and such are still self-managing as to Z-order. These issues may be slightly more obvious in xmonad, but will show in others sometimes.

**Comment by IvanMalison** [comment]
@geekosaur This isn't about losing docks on restart.

What's wrong with the idea that sgf-dma suggested of lowering docks before they are unmanaged.

> because docks and such are still self-managing as to Z-order.

What exactly is meant by this?

**Comment by geekosaur** [comment]
I appear to have crossed this with another ticket in the recent barrage of comments. (That one being deeply confused as to what's doing what.)

Window managers don't tell docks where and how to draw; docks tell window managers how to avoid whatever they did. But don't include which windows are to be considered "on top". EWMH's layers are supposed to avoid collisions there, but are at most advisory and don't work if clients don't specifically check or make incorrect assumptions. I see bugs in this area in other window managers as well.

**Comment by sgf-dma** [comment]
Hi.

Sorry for disappearing for so long time, i haven't any free time to look into
this.

[Here is the code](https://gist.github.com/sgf-dma/b821198180e12a9f71e84660a3b98c11) demonstrating the problem, as i understand it.

    import XMonad
    import XMonad.Operations
    import XMonad.Util.EZConfig
    import XMonad.Util.DebugWindow
    import XMonad.Hooks.DebugStack
    import XMonad.Hooks.ManageDocks
    import qualified XMonad.StackSet as W
    import XMonad.Hooks.EwmhDesktops

    import Control.Monad
    import Control.Exception
    import Numeric
    import Data.List

    main :: IO ()
    main = do
            let xcf = docks $ def
                        { modMask = mod4Mask
                        , logHook = debugSt >> logH >> logHook def
                        , layoutHook = avoidStruts (layoutHook def)
                        , handleEventHook = handleEventHook def <+> fullscreenEventHook
                        }
                        `additionalKeys`
                        [ ((mod4Mask, xK_b), sendMessage ToggleStruts)
                        , ((mod4Mask, xK_x), spawn "xmobar ~/.xmobarrcAlt")
                        , ((mod4Mask, xK_m), spawn "chromium")
                        , ((mod4Mask, xK_r), refresh)
                        ]
            xmonad xcf

    -- Change order of `debugStack` output to natural..
    debugSt :: X ()
    debugSt = debugStackString >>= trace . unlines . reverse . lines

    logH :: X ()
    logH    = withWindowSet $ \ws -> do
        trace "Tree..."
        d <- asks display
        r <- asks theRoot
        (_, _, ts) <- liftIO $ queryTree d r
        ls <- mapM debugWindow ts
        trace (intercalate "\n" ls)
      where
        showWs :: [Window] -> ShowS
        showWs      = fmap (intercalate "\n") . mapM showHex

After line `Tree..` the content of X server window stack (`queryTree`) will be
shown.

First i start chromium:

    * 01200001 "New Tab - Chromium" 1918x1078+1@0,0 chromium-browser/Chromium-browser (25563)@reiji.sgf.un
    Workspace "1" (current)

    Tree...
    [00400001 10x10@-20,-20]
    [00a00001 "ibus-x11" 10x10@10,10 ibus-x11/Ibus-x11  "ibus-x11"(25526)@reiji.sgf.un]
    [00a00003 "ibus-xim" 1x1@0,0 (25526)@reiji.sgf.un]
    [00c00001 "ibus-ui-gtk3" 10x10@10,10 ibus-ui-gtk3/Ibus-ui-gtk3  "ibus-ui-gtk3"(25519)@reiji.sgf.un]
    [00c00007 "IBus Panel" 200x200@0,0 ibus-ui-gtk/ibus-ui-gtk (25519)@reiji.sgf.un]
    [01000001 "chromium" 10x10@10,10 chromium/Chromium  "chromium"(25563)@reiji.sgf.un]
    [01000003 "chromium" 200x200@0,0 chromium/Chromium (25563)@reiji.sgf.un]
    [0100001e "chromium" 200x200@0,0 chromium/Chromium (25563)@reiji.sgf.un]
    01200001 "New Tab - Chromium" 1918x1078+1@0,0 chromium-browser/Chromium-browser (25563)@reiji.sgf.un
    [01200005 1x1@0,0]

then i start `xmobar`

    * 01200001 "New Tab - Chromium" 1918x1078+1@0,0 chromium-browser/Chromium-browser (25563)@reiji.sgf.un
    Workspace "1" (current)

    Tree...
    [00400001 10x10@-20,-20]
    [00a00001 "ibus-x11" 10x10@10,10 ibus-x11/Ibus-x11  "ibus-x11"(25526)@reiji.sgf.un]
    [00a00003 "ibus-xim" 1x1@0,0 (25526)@reiji.sgf.un]
    [00c00001 "ibus-ui-gtk3" 10x10@10,10 ibus-ui-gtk3/Ibus-ui-gtk3  "ibus-ui-gtk3"(25519)@reiji.sgf.un]
    [00c00007 "IBus Panel" 200x200@0,0 ibus-ui-gtk/ibus-ui-gtk (25519)@reiji.sgf.un]
    [01000001 "chromium" 10x10@10,10 chromium/Chromium  "chromium"(25563)@reiji.sgf.un]
    [01000003 "chromium" 200x200@0,0 chromium/Chromium (25563)@reiji.sgf.un]
    [0100001e "chromium" 200x200@0,0 chromium/Chromium (25563)@reiji.sgf.un]
    01200001 "New Tab - Chromium" 1918x1078+1@0,0 chromium-browser/Chromium-browser (25563)@reiji.sgf.un
    [01200005 1x1@0,0]
    [01600001 1x1@0,0]
    [01400006 1x1@0,0]
    01800002 "xmobar" 1920x19@0,1061 xmobar (25662)

as you can see, `xmobar` is _higher_ in X server stack, than chromium. But
it's already ignored by `xmonad`, so `restackWindows` run in `windows` can't
place chromium on top of it. In this situation full screen chromium window
will go _under_ xmobar. I.e. this bug occurs.

Then i start another new window (`xterm`).

    * 01c0000f "xterm" 958x1059+1@0,0 xterm/XTerm  "xterm"(25684)@reiji.sgf.un
      01200001 "New Tab - Chromium" 958x1059+1@960,0 chromium-browser/Chromium-browser (25563)@reiji.sgf.un
    Workspace "1" (current)

    Tree...
    [00400001 10x10@-20,-20]
    [00a00001 "ibus-x11" 10x10@10,10 ibus-x11/Ibus-x11  "ibus-x11"(25526)@reiji.sgf.un]
    [00a00003 "ibus-xim" 1x1@0,0 (25526)@reiji.sgf.un]
    [00c00001 "ibus-ui-gtk3" 10x10@10,10 ibus-ui-gtk3/Ibus-ui-gtk3  "ibus-ui-gtk3"(25519)@reiji.sgf.un]
    [00c00007 "IBus Panel" 200x200@0,0 ibus-ui-gtk/ibus-ui-gtk (25519)@reiji.sgf.un]
    [01000001 "chromium" 10x10@10,10 chromium/Chromium  "chromium"(25563)@reiji.sgf.un]
    [01000003 "chromium" 200x200@0,0 chromium/Chromium (25563)@reiji.sgf.un]
    [0100001e "chromium" 200x200@0,0 chromium/Chromium (25563)@reiji.sgf.un]
    [01200005 1x1@0,0]
    [01600001 1x1@0,0]
    [01400006 1x1@0,0]
    01800002 "xmobar" 1920x19@0,1061 xmobar (25662)
    01200001 "New Tab - Chromium" 958x1059+1@960,0 chromium-browser/Chromium-browser (25563)@reiji.sgf.un
    01c0000f "xterm" 958x1059+1@0,0 xterm/XTerm  "xterm"(25684)@reiji.sgf.un
    [01c0001c 1x1+1@0,0]
    [00a00005 1x1+1@0,0]

Because `xterm` window is mapped after `xmobar` window, it appears on top of
it in X server stack. Thus, running `restackWindows` will _bring up_ chromium
window on top of `xmobar` too. Now fullscreen chromium window will go _over_
xmobar, as it should. But this workaround will not always work. It relies on the
order of windows in a list `rs` returned by `runLayout` in `windows`
(assuming, there are no floating windows). Particularly, considering Tall
layout, new window must be leftmost (last element in `up`) for this fix to
work, i suppose.

Regarding the fix, i don't yet try to implement it, but using `lowerWindow` on
all docks before unmanaging them is probably the wrong choice, because then a
dock may go under desktop (i.e. when file manager used to manage desktop). I
think, we may restack a new dock below all managed windows before unmanaging
its, or keep a list of all docks and restack them together below all managed
windows..

**Comment by geekosaur** [comment]
I'll look this over later, but `debugStack` doesn't involve `XQueryTree`; it's reporting the `Stack` associated with the current or all workspaces, from the current `windowset`.

**Comment by sgf-dma** [comment]
And [here](https://gist.github.com/sgf-dma/33d18ed193c30e5f211ce273dd73f49a) is proof of concept fix:

    import XMonad
    import XMonad.Util.EZConfig
    import XMonad.Util.DebugWindow
    import XMonad.Hooks.DebugStack
    import XMonad.Hooks.ManageDocks
    import qualified XMonad.StackSet as W
    import XMonad.Hooks.EwmhDesktops

    import Data.List

    main :: IO ()
    main = do
            let xcf = def
                        { modMask = mod4Mask
                        , layoutHook = avoidStruts (layoutHook def)
                        -- Lower dock before unmanaging. The position of
                        -- `lowerDock` in `ManageHook` does not matter, because it
                        -- only runs X actions.
                        , manageHook = lowerDock <+> manageDocks <+> manageHook def
                        , handleEventHook = fullscreenEventHook <+> docksEventHook <+> handleEventHook def
                        , startupHook = docksStartupHook <+> startupHook def
                        }
                        `additionalKeys`
                        [ ((mod4Mask, xK_b), sendMessage ToggleStruts)
                        , ((mod4Mask, xK_x), spawn "xmobar ~/.xmobarrcAlt")
                        , ((mod4Mask, xK_m), spawn "chromium")
                        , ((mod4Mask, xK_r), refresh)
                        ]
            xmonad xcf

    -- Change order of `debugStack` output to natural..
    debugSt :: X ()
    debugSt = debugStackString >>= trace . unlines . reverse . lines

    logH :: X ()
    logH    = do
        trace "Tree..."
        d <- asks display
        r <- asks theRoot
        (_, _, ts) <- liftIO $ queryTree d r
        ls <- mapM debugWindow ts
        trace (intercalate "\n" ls)

    -- | Restack dock under lowest managed window.
    lowerDock :: ManageHook
    lowerDock = checkDock --> do
        w <- ask
        mlw <- liftX $ findLowest
        case mlw of
          Just lw   -> liftX $ do
            d <- asks display
            s <- debugWindow lw
            trace "_Before_ restacking dock:"
            logH
            trace $ "Lowest managed window: " ++ s
            liftIO $ restackWindows d [lw, w]
            trace "_After_ restack dock:"
            logH
            return idHook
          Nothing   -> return idHook

    -- | Find lowest managed window.
    findLowest :: X (Maybe Window)
    findLowest  = withWindowSet $ \ws -> do
        d <- asks display
        r <- asks theRoot
        (_, _, ts) <- liftIO $ queryTree d r
        return (find (`W.member` ws) ts)

This fix (`lowerDock` function) `restack`-s dock below _lowest_ window
managed by xmonad _before_ it'll become unmanaged (by `doIgnore` in
`manageDocks`). For now, i just rewrite `docks` as separate three functions
and add `lowerDock` before `manageDocks`.

@IvanMalison , @dkasak , @de-vri-es , please, try this. You need to unwrap `docks`, add `lowerDock` to
`manageHook` and copy `lowerDock` and `findLowest` functions.

**Comment by geekosaur** [comment]
`queryTree` is a rather expensive function; if possible, try to find a way to do it using the existing `Stack`.

**Comment by dkasak** [comment]
@sgf-dma I just tried your POC and I confirm that it fixes the issue.

**Comment by sgf-dma** [comment]
@geekosaur I don't think this is possible with `Stack`, because it just
doesn't have needed information.

The window order is affected by program start order and used layout (`windows`
runs `restackWindows` with a list generated by `runLayout`) and floating
windows order. But considering that each workspace has different floating
windows and may have different layout, i don't know how we can predict window
order in X stack. And even if we can, i think, it'll be both very complex and
unreliable.

There're two simple options, though:

1. `lowerWindow`. It does not require to run `queryTree`, but with this
   approach we can easily end up with dock placed under desktop (e.g. under
   nautilus window managing desktop). In theory, it's possible to workaround
   this by keeping track in xmonad of all dock windows (including desktop) and
   restack them in correct order after lowering one of them. Whether this is
   complex or not, it'll require from the user to start all docks after xmonad
   has started, so we can detect them in `ManageHook` and record somewhere.

2. `restackWindows`. The one, i've implemented. It's far simpler, than first,
   but requires `queryTree` for finding lowest window in X stack. But we run
   this only _once_ for each _new_ dock. I.e. this will be 2-3 times per whole
   user session. I don't think this will create any performance problems.

Managing all docks within xmonad (the 1st approach) also has some advantages:
we may enforce correct dock order and overlapping in X stack.  And e.g. even
specify some "weight" for docks, which should affect dock placement during
`restackWindows`.. but, i think, this is too much just for fixing this issue.

**Comment by geekosaur** [comment]
You may need to duplicate `XMonad.Operations.windows` to do it, but note that that does `XRestackWindows` with an order generated from the `Stack`. (Which includes floating windows — which is why they behave so obnoxiously. The tiled window list is generated by an application of (\\) to remove the contents of `floating`.) So `windows` always restacks all visible, managed windows, and duplicating its logic will give you the full stacking order.

This does miss things like desktop windows, but crawling up the stack checking window props is much lighter weight than `XQueryTree`. Or just checking if the window is in the `StackSet`, which also avoids that X11 roundtrip.

At X11 protocol level, `XQueryTree` does not return an array; it's a separate X11 event per returned window, plus some extra "overhead" events. Go look at how it's implemented in `xcb` which exposes the protocol level to C instead of hiding the individual events like `xlib` does. The whole point of the `ManageDocks` redesign in 0.12 was to get rid of doing this every time the layout was run, because with more than a handful of mapped windows it's visibly very slow; but you're bringing that back again. If we really do have to `XQueryTree` every time, then just revert `ManageDocks` back to 0.11 where it was simple and reliable and as slow as it will be with this change.

BTW, there actually is a reason we might want to go ahead and partially manage dock and desktop windows: proper EWMH layer support requires it. (We currently fake it, and there are cases when the desktop window can float over other windows as a result, granting that this usually requires whatever opened the desktop window to have certain bugs.) It would also provide the necessary hooks for better float management at some point, since adding a proper floating layer would be almost trivial once layers are there for desktop and dock windows.

**Comment by geekosaur** [comment]
Well, the above is actually overstating a bit because 0.11 also did two property queries for every mapped window every time the layout was run. I think this can still be avoided, but possibly not since changing it to only check windows not in the `Stack` is part of what caused the subsequent `dzen` breakage. (More precisely, it was being treated as a flxed-size window and marked as a float instead of a dock, because it didn't have the `_NET_WINDOW_TYPE_DOCK` property when initially mapped, IIRC. This worked before because it was in effect treated as a `_NET_WINDOW_STATE_ABOVE` dock, as long as it didn't accept focus.)

This all ends up being expensive to do properly unless you bite the bullet and add layer support so you can properly place windows in layers as part of restacking. Or switch to `raiseWindow` / `lowerWindow` instead of `restackWindows`, but that will require even more window accounting we don't do and will require a lot more support to get right, on top of requiring a complete redesign of `X.O.windows`. (And make layer support mandatory instead of letting us fake it as now.)

**Comment by sgf-dma** [comment]
> You may need to duplicate XMonad.Operations.windows to do it, but note that
> that does XRestackWindows with an order generated from the Stack. (Which
> includes floating windows — which is why they behave so obnoxiously. The
> tiled window list is generated by an application of (\) to remove the
> contents of floating.) So windows always restacks all visible, managed
> windows, and duplicating its logic will give you the full stacking order.

I don't think so. There're unmapped windows on other (non visible workspaces),
which are not restacked at `windows` run. Even if i copy the `windows` logic
here, i only can find lowest _now visible_ window. It may not be the lowest
_managed_ by xmonad window (i.e. some windows on non-visible workspace may be
lower in X stack).

> The whole point of the ManageDocks redesign in 0.12 was to get rid of doing
> this every time the layout was run, because with more than a handful of
> mapped windows it's visibly very slow; but you're bringing that back again.
> If we really do have to XQueryTree every time, then just revert ManageDocks
> back to 0.11 where it was simple and reliable and as slow as it will be with
> this change.

I don't understand your point. In my code `queryTree` runs _only_, when new
dock is started (it's in `ManageHook`, guarded by `checkDock`). And if you are
not killing and starting docks every second, that will be 2-3 (may be 5) times
per _whole desktop session_. How this can affect `runLayout` ?

> BTW, there actually is a reason we might want to go ahead and partially
> manage dock and desktop windows: proper EWMH layer support requires it. (We
> currently fake it, and there are cases when the desktop window can float
> over other windows as a result, granting that this usually requires whatever
> opened the desktop window to have certain bugs.) It would also provide the
> necessary hooks for better float management at some point, since adding a
> proper floating layer would be almost trivial once layers are there for
> desktop and dock windows.

This is a reason for going by way 2? It sounds to me, like some early draft
for very-very distant future, especially considering number of open issues and
pull requests. If this issue will wait for all that to be implemented, it'll be
open forever.

**Comment by sgf-dma** [comment]
> Well, the above is actually overstating a bit because 0.11 also did two
> property queries for every mapped window every time the layout was run.

I've looked in 0.11 code. Well, it runs `queryTree` in `caclGap`, which was
used in `modifyLayout`, but this has absolutely nothing in common with how i
use it here.

> I think this can still be avoided, but possibly not since changing it to
> only check windows not in the Stack is part of what caused the subsequent
> dzen breakage. (More precisely, it was being treated as a flxed-size window
> and marked as a float instead of a dock, because it didn't have the
> _NET_WINDOW_TYPE_DOCK property when initially mapped, IIRC. This worked
> before because it was in effect treated as a _NET_WINDOW_STATE_ABOVE dock,
> as long as it didn't accept focus.)

Is there a bug report about this? (to read full details)

By the way, [here](https://gist.github.com/sgf-dma/d4f623277dc47df6a1ef7e3dbbcb321f) is proof of concept of another approach (with `lowerWindow`).


    import XMonad
    import XMonad.Util.EZConfig
    import XMonad.Util.DebugWindow
    import XMonad.Hooks.DebugStack
    import XMonad.Hooks.ManageDocks
    import XMonad.Hooks.EwmhDesktops
    import XMonad.Hooks.ManageHelpers
    import qualified XMonad.Util.ExtensibleState as XS
    import XMonad.Util.WindowProperties (getProp32s)

    import Data.List
    import Data.Monoid

    main :: IO ()
    main = do
            let xcf = def
                        { modMask = mod4Mask
                        , layoutHook = avoidStruts (layoutHook def)
                        , manageHook = addDock <+> manageDocks <+> manageHook def
                        , handleEventHook = delDock <+> fullscreenEventHook <+> docksEventHook <+> handleEventHook def
                        , startupHook = docksStartupHook <+> startupHook def
                        , logHook = logDL >> logHook def
                        }
                        `additionalKeys`
                        [ ((mod4Mask, xK_b), sendMessage ToggleStruts)
                        , ((mod4Mask, xK_x), spawn "xmobar ~/.xmobarrcAlt")
                        , ((mod4Mask, xK_m), spawn "chromium")
                        , ((mod4Mask, xK_r), refresh)
                        ]
            xmonad xcf

    -- Change order of `debugStack` output to natural..
    debugSt :: X ()
    debugSt = debugStackString >>= trace . unlines . reverse . lines

    logH :: X ()
    logH    = do
        trace "Tree..."
        d <- asks display
        r <- asks theRoot
        (_, _, ts) <- liftIO $ queryTree d r
        ls <- mapM debugWindow ts
        trace (intercalate "\n" ls)

    data DockT  = DockA Window
                | DesktopA Window
      deriving (Show, Eq)

    checkDesktop :: Query Bool
    checkDesktop = ask >>= \w -> liftX $ do
        desk <- getAtom "_NET_WM_WINDOW_TYPE_DESKTOP"
        mbr <- getProp32s "_NET_WM_WINDOW_TYPE" w
        case mbr of
            Just rs -> return $ any (== desk) (map fromIntegral rs)
            _       -> return False

    newtype DockList    = DockList [DockT]
      deriving (Show)

    findDock :: Window -> [DockT] -> Maybe DockT
    findDock w    = find go
      where
        go :: DockT -> Bool
        go (DockA x)    = w == x
        go (DesktopA x) = w == x

    getPanels :: DockList -> [DockT]
    getPanels (DockList ds) = filter go ds
      where
        go (DockA _)    = True
        go _            = False

    getDesktops :: DockList -> [DockT]
    getDesktops (DockList ds) = filter go ds
      where
        go (DesktopA _) = True
        go _            = False

    unDock :: DockT -> Window
    unDock (DockA x)    = x
    unDock (DesktopA x) = x

    instance ExtensionClass DockList where
        initialValue = DockList []


    addDesktop :: MaybeManageHook
    addDesktop    = checkDesktop  -?> do
        w <- ask
        liftX $ XS.modify (\(DockList zs) -> DockList (DesktopA w : zs))
        return idHook

    addPanel :: MaybeManageHook
    addPanel    = checkDock  -?> do
        w <- ask
        liftX $ XS.modify (\(DockList zs) -> DockList (DockA w : zs))
        return idHook

    addDock :: ManageHook
    addDock   = checkDock --> do
        composeOne [addDesktop, addPanel]
        liftX $ do
          dl <- XS.get
          let ds = getDesktops dl
              ps = getPanels dl
              rs = map unDock (ps ++ ds)
          d <- asks display
          trace "_Before_ lowering dock:"
          logH
          liftIO $ do
            lowerWindow d (head rs)
            restackWindows d rs
          trace "_After_ lowering dock:"
          logH
          return mempty

    delDock :: Event -> X All
    delDock (DestroyWindowEvent {ev_window = w}) = do
        DockList ds <- XS.get
        let md = w `findDock` ds
        case md of
          Just d    -> XS.put (DockList (d `delete` ds)) >> return mempty
          Nothing   -> return mempty
    delDock _ = return mempty

    logDL :: X ()
    logDL   = do
        DockList ds <- XS.get
        trace $ "Docks list: " ++ intercalate ", " (map show ds)


@IvanMalison , @dkasak , @de-vri-es , please, try this too. You need to unwrap
`docks`, add `addDock` to `manageHook` and `delDock` to `handleEventHook`.

@geekosaur , so what do you think about this?

**Comment by sureyeaah** [comment]
Is there any update on this?

**Comment by liskin** [comment]
A bit weird that nobody in this discussion mentioned that xmobar can lower itself on startup if configured to do so, and so can trayer. I'd be surprised if other docks couldn't do that as well.

That being said, if one wishes to run several such docks over another, having them all lower themselves may result in random order depending which one lower themselves first. And in that case it makes sense to get xmonad involved. In my case, that's putting xmobars below trayer: https://github.com/liskin/dotfiles/commit/3d19988009fcc3f6c05c19f30e748d3e65df3e6d

But I'm not sure if this is something general enough that xmonad could do this itself. How would it know if the user prefers trayer above xmobar or the other way around? How would it know that dockxyz is a desktop and dockabc is a panel? We'd need a module that makes the desired order configurable somehow.

Anyway, I agree with @sgf-dma here that using `queryTree` is fine as long as it's done only when the dock appears (or, in the case of my config, only when the dock lowers itself).

**Comment by liskin** [comment]
I think the hack that was merged in #490 can be abused to catch panels lowering themselves below the desktop, so maybe this finally has a "solution"?

Another related PR is #492 which makes it safer to let xmobar use override redirect and lower itself, instead of marking itself as a dock.

**Comment by slotThe** [comment]
> I think the hack that was merged in #490 can be abused to catch panels lowering themselves below the desktop, so maybe this finally has a "solution"?

I would agree.  If anyone still has any open issues here feel free to chime in though

**Comment by dkasak** [comment]
Is this supposed to work for all windows or just for new windows which were created after an xmonad restart? I just updated and while new windows behave as expected, old windows still appear under my docks.

**Comment by real-fht** [comment]
Hi, I know this might be late, but here's my "solution" to this issue.

```hs

-- Note: replace "..." with other imports you may need.
import XMonad.Hooks.ManageHelpers (..., doLower)
import XMonad.Hooks.ManageDocks   (..., checkDock)

-- And then add this to your manageHook
myManageHook = composeAll 
  [
  -- ...Other stuff you have here...
  checkDock                   --> doLower
  ]

```

**Comment by sverona** [comment]
> Hi, I know this might be late, but here's my "solution" to this issue.
> 
> ```haskell
> -- Note: replace "..." with other imports you may need.
> import XMonad.Hooks.ManageHelpers (..., doLower)
> import XMonad.Hooks.ManageDocks   (..., checkDock)
> 
> -- And then add this to your manageHook
> myManageHook = composeAll 
>   [
>   -- ...Other stuff you have here...
>   checkDock                   --> doLower
>   ]
> ```

It should be noted that `doLower` doesn't exist prior to xmonad-contrib 0.17. This bit me on Arch, on which [xmonad{,-contrib} is unmaintained since Goddess only knows when.](https://bbs.archlinux.org/viewtopic.php?id=274904)

**Comment by real-fht** [comment]
@sverona Well based on the official [install tutorial](https://xmonad.org/INSTALL.md), it's preferable that you install xmonad from source on distros like Arch that statically links haskell libraries.

Nonetheless, yeah the `doLower` function is only available on 0.17 and over.
